### DISCUSSION TITLE
Blue light does not come on when copying bitstream and RedPitaya crashes when trying to run a sequence

### DISCUSSION THREAD

**Discussion by LUMC-LowFieldMRI** [vnegnev/marcos_extras#9] [Debugging]
The setup_marcos.sh script runs to completion but the blue light does not come on when it is supposed to come on during the installation. When we try to run a pulse sequence the Red Pitaya crashes and we have to unplug it to get in to a workable state, do you know what could be causing this?

The output of the script is:
```Setting up MaRCoS on IP 169.254.14.10...
Setting date on the STEMlab based on the host date...
date: invalid date 'Lw2 18 Oki 2022 14:22:50 UTC'
Copying FPGA bitstream...
unknown option -- O
usage: scp [-346BCpqrTv] [-c cipher] [-F ssh_config] [-i identity_file]
            [-J destination] [-l limit] [-o ssh_option] [-P port]
            [-S program] source ... target
unknown option -- O
usage: scp [-346BCpqrTv] [-c cipher] [-F ssh_config] [-i identity_file]
            [-J destination] [-l limit] [-o ssh_option] [-P port]
            [-S program] source ... target
Writing bitstream to FPGA, ocra Linux image
(you should see a blue light appear) ...
Pseudo-terminal will not be allocated because stdin is not a terminal.
Killing any existing server instances...
killall: marcos_server: no process killed
Copying MaRCoS server...
Cloning into '/tmp/marcos_server'...
remote: Enumerating objects: 16, done.
remote: Counting objects: 100% (16/16), done.
remote: Compressing objects: 100% (15/15), done.
remote: Total 16 (delta 0), reused 4 (delta 0), pack-reused 0
Unpacking objects: 100% (16/16), 95.85 KiB | 434.00 KiB/s, done.
unknown option -- O
usage: scp [-346BCpqrTv] [-c cipher] [-F ssh_config] [-i identity_file]
            [-J destination] [-l limit] [-o ssh_option] [-P port]
            [-S program] source ... target
Compiling MaRCoS server on the STEMlab...
Pseudo-terminal will not be allocated because stdin is not a terminal.
CMake Error: The source directory "/tmp/marcos_server/src" does not exist.
Specify --help for usage, or press the help button on the CMake GUI.
make: *** No targets specified and no makefile found.  Stop.
cp: cannot stat 'marcos_server': No such file or directory
You can run the MaRCoS server by entering the following command:
ssh [root@169.254.14.10](mailto:root@169.254.14.10) "~/marcos_server"
```
[VN: edited to print dump as a block]

**Comment by vnegnev** [comment]
Looks like a few things are going wrong with the script:
- The date isn't in the standard format - this will need a change to the script on Line 18 (or you can set the SDRLab's date manually by SSHing in)
- The scp command doesn't recognise the `-O` flag that the scripts are using - this is probably because the scp version in your desktop environment is old. I updated the command quite recently though, so the older script will probably work - I probably need to make the shell script more sophisticated and check the version of SCP on the host machine to avoid this problem.

Can you checkout the script from the previous commit (83cbde59eea1...) and see if the blue light turns on then?

**Reply by vnegnev** [reply]
Looks like I should update the copy protocol to SFTP in general, since SCP is considered outdated. I'll do this in a few days.

**Comment by LUMC-LowFieldMRI** [comment]
Hi Vlad, we ran the old version and fixed the date issue (maybe good to add to future versions of the marcos setup script (LANG=en_us_88591; date -u) instead of just (date -u), we also had the same issue on our system) but the blue light still doesn't come on. The output of the script is was:

```Setting up MaRCoS on IP 169.254.14.10...
Setting date on the STEMlab based on the host date...
Wed Oct 19 07:41:44 UTC 2022
Copying FPGA bitstream...
ocra_mri_rp-122.bit.bin                                                                                                                                                          100% 3951KB  10.0MB/s   00:00    
ocra_mri_rp-122.dtbo                                                                                                                                                             100% 4751     2.8MB/s   00:00    
Writing bitstream to FPGA, ocra Linux image
(you should see a blue light appear) ...
Pseudo-terminal will not be allocated because stdin is not a terminal.
Killing any existing server instances...
killall: marcos_server: no process killed
Copying MaRCoS server...
Cloning into '/tmp/marcos_server'...
remote: Enumerating objects: 16, done.
remote: Counting objects: 100% (16/16), done.
remote: Compressing objects: 100% (15/15), done.
remote: Total 16 (delta 0), reused 4 (delta 0), pack-reused 0
Unpacking objects: 100% (16/16), 95.85 KiB | 57.00 KiB/s, done.
mkdir: cannot create directory '/tmp/marcos_server': File exists
[readme.org](http://readme.org/)                                                                                                                                                                       100%  599   270.4KB/s   00:00    
iface.cpp                                                                                                                                                                        100% 9704     5.3MB/s   00:00    
hardware.cpp                                                                                                                                                                     100%   25KB   4.4MB/s   00:00    
iface.hpp                                                                                                                                                                        100% 5232     4.0MB/s   00:00    
marcos_server.cpp                                                                                                                                                                100% 1269     1.5MB/s   00:00    
CMakeLists.txt                                                                                                                                                                   100%  771     1.1MB/s   00:00    
server_config.h                                                                                                                                                                  100%  897     1.1MB/s   00:00    
version.hpp                                                                                                                                                                      100%  217   368.9KB/s   00:00    
mpack.c                                                                                                                                                                          100%  211KB  10.0MB/s   00:00    
mpack.h                                                                                                                                                                          100%  224KB  10.3MB/s   00:00    
hardware.hpp                                                                                                                                                                     100% 3422     2.9MB/s   00:00    
Compiling MaRCoS server on the STEMlab...
Pseudo-terminal will not be allocated because stdin is not a terminal.
mkdir: cannot create directory 'build': File exists
-- Configuring done
-- Generating done
-- Build files have been written to: /tmp/marcos_server/build
Scanning dependencies of target marcos_server
[ 40%] Building CXX object CMakeFiles/marcos_server.dir/marcos_server.cpp.o
[ 40%] Building CXX object CMakeFiles/marcos_server.dir/hardware.cpp.o
/tmp/marcos_server/src/hardware.cpp: In member function 'int hardware::run_request(server_action&)':
/tmp/marcos_server/src/hardware.cpp:322:11: warning: unused variable 'k' [-Wunused-variable]
  322 |      auto k = (char *)hw_memcpy(_flo_mem + local_mem_offset,
      |           ^
/tmp/marcos_server/src/hardware.cpp:41:7: warning: variable 'hr' set but not used [-Wunused-but-set-variable]
   41 |  auto hr = sa.get_command_and_start_reply("halt_and_reset", status);
      |       ^~
/tmp/marcos_server/src/hardware.cpp:52:7: warning: variable 'rm' set but not used [-Wunused-but-set-variable]
   52 |  auto rm = sa.get_command_and_start_reply("read_mem", status);
      |       ^~
/tmp/marcos_server/src/hardware.cpp:154:7: warning: variable 'rr' set but not used [-Wunused-but-set-variable]
  154 |  auto rr = sa.get_command_and_start_reply("read_rx", status);
      |       ^~
[ 60%] Building CXX object CMakeFiles/marcos_server.dir/iface.cpp.o
[ 80%] Building C object CMakeFiles/marcos_server.dir/mpack/mpack.c.o
[100%] Linking CXX executable marcos_server
[100%] Built target marcos_server
You can run the MaRCoS server by entering the following command:
ssh [root@169.254.14.10](mailto:root@169.254.14.10) "~/marcos_server"
```

**Comment by vnegnev** [comment]
That's very strange - copying the FPGA bitstream is handled by the https://github.com/vnegnev/marcos_extras/blob/master/copy_bitstream.sh script.

You can manually copy over the bin and dtbo files to `/lib/firmware` on the SDRLab, then load them onto the FPGA through an SSH terminal. The commands to run on the SDRLab are from Line 29 to Line 38 in `copy_bitstream.sh`. Hopefully you see an error or other debugging information from them - I'll try it on my SDRLab and see if there's a problem tonight as well.

**Comment by vnegnev** [comment]
So I've tried the commands in `copy_bitstream.sh` again, and on my SDRLab they work fine. Maybe it's worth making sure that the bitstream and dtbo files are uncorrupted on the SDRLab's filesystem in your case. You can run the following to make sure:

```
root@redpitaya:~# cd /lib/firmware
root@redpitaya:/lib/firmware# ls -al
total 3968
drwxr-xr-x 2 root root    4096 May 18 18:16 .
drwxr-xr-x 8 root root    4096 May 18 18:16 ..
-rw-r--r-- 1 root root 4045568 May 18 18:16 ocra_mri.bit.bin
-rw-r--r-- 1 root root    4751 May 18 18:16 ocra_mri.dtbo
root@redpitaya:/lib/firmware# sha1sum *
222ce999cfea103946e7dc1551431496fcb69c41  ocra_mri.bit.bin
d099054315aec10362849860c307082972163882  ocra_mri.dtbo
root@redpitaya:/lib/firmware# 
```

The sha1sum values should match mine exactly (you can also run `sha1sum` against the files in the `marcos_extras` repo to make sure you have the same files as me).

To load the bitstream on, you can run the following commands manually:

```
root@redpitaya:/lib/firmware# echo 0 > /sys/class/fpga_manager/fpga0/flags
root@redpitaya:/lib/firmware# mkdir /sys/kernel/config/device-tree/overlays/full
root@redpitaya:/lib/firmware# echo -n "ocra_mri.dtbo" > /sys/kernel/config/device-tree/overlays/full/path
```

After the last command, the blue LED should immediately turn on. None of the commands print any messages. If you have the right files in the `/lib/firmware` folder and these commands ran fine but the LED still didn't turn on, then either a) there's something wrong with your SD card image and it's worth preparing a new one, or b) your SDRLab is somehow faulty. This is really unlikely though.

The other thing to check is whether you're using a decent power supply (5V, 2A minimum) for the SDRLab? I've seen intermittent problems with lower-rated supplies.

Also, just to be safe, make sure you're always using the `poweroff` shell command over SSH when you shut the SDRLab down, to avoid data corruption on the SD card. (Of course it's somewhat unavoidable when the server is run without a loaded bitstream since the system locks up, but mentioning it just in case.)

Hope that helps!

**Reply by aaristotlerobert** [reply]
Hi Vlad
I am robert and part of the LowFieldMRI group. I actually ran What you said we should do. But when i ran **root@redpitaya:/lib/firmware# ls -al** at the SDRlab the next output was
```total 3980
drwxr-xr-x 2 root root    4096 Apr  2 23:27 .
drwxr-xr-x 7 root root    4096 Apr  2 23:18 ..
-rw-r--r-- 1 root root 4045568 Apr  2 23:28 ocra_mri.bit.bin
-rw-r--r-- 1 root root    4751 Apr  2 23:28 ocra_mri.dtbo
-rw-r--r-- 1 root root    4248 Mar 10 11:45 regulatory.db
-rw-r--r-- 1 root root    1182 Mar 10 11:45 regulatory.db.p7s
```
the first line is different from yours according to your output when you ran same command. This is because mine it is  `total 3980` yet yours is `total 3968`  I have two additional lines that is `-rw-r--r-- 1 root root    4248 Mar 10 11:45 regulatory.db` and `-rw-r--r-- 1 root root    1182 Mar 10 11:45 regulatory.db.p7s` of which you didn't have; could they be an issue.

Furthermore, when I ran `root@redpitaya:/lib/firmware# sha1sum *` I got the following output 
```222ce999cfea103946e7dc1551431496fcb69c41  ocra_mri.bit.bin
d099054315aec10362849860c307082972163882  ocra_mri.dtbo
d47470035bb627219c2cc4c585bfe9de4baacc17  regulatory.db
e673f2f721456eeb98258c900a3b881fc138a0ac  regulatory.db.p7s
```
the first two lines are same as you got but I got two more other lines that is  `d47470035bb627219c2cc4c585bfe9de4baacc17  regulatory.db` and `e673f2f721456eeb98258c900a3b881fc138a0ac  regulatory.db.p7s`  again are they necessary?

The following were the next commands with their corresponding output scripts

```root@redpitaya:/lib/firmware# echo 0 > /sys/class/fpga_manager/fpga0/flags                        
root@redpitaya:/lib/firmware# mkdir /sys/kernel/config/device-tree/overlays/full
mkdir: cannot create directory '/sys/kernel/config/device-tree/overlays/full': File exists
root@redpitaya:/lib/firmware# echo -n "ocra_mri.dtbo" > /sys/kernel/config/device-tree/overlays/full/path
-sh: echo: write error: Operation not permitted
root@redpitaya:/lib/firmware# mkdir /sys/kernel/config/device-tree/overlays/full
mkdir: cannot create directory '/sys/kernel/config/device-tree/overlays/full': File exists
```
At `root@redpitaya:/lib/firmware# echo 0 > /sys/class/fpga_manager/fpga0/flags ` it seemed happy but at `root@redpitaya:/lib/firmware# mkdir /sys/kernel/config/device-tree/overlays/full` and the following commands, it did not workout as expected. I think where the issue is. 
**Do you have any further suggestions?** 
Thanks.

[VN: edited block formatting]

**Reply by vnegnev** [reply]
Hi Robert, welcome to the project!

The `regulatory.db` and `regulatory.db.p7s` files shouldn't matter - I had them when I ran the test as well but I deleted them because I thought they weren't in the SD card image. I guess they were! Either way they're fine.

I think the first line of the `sha1sum *` command's output got cut off (I'd have expected to see the value for `ocra_mri.bit.bin`), but if your SHA-1 sum of `ocra_mri.bit.bin` matched mine, that means you definitely have the correct files present.

I'm concerned about the errors you're seeing for the next commands - these are the main problem. The fact that they don't run properly means you probably have something going wrong with your boot process, or the SD card is corrupted. I'll prepare the boot log from my SDRLab this evening, then we can compare. Can you format a spare blank SD card with the Linux image as described in Step 1.2 here: https://github.com/vnegnev/marcos_extras/wiki/guide_setting_marcos_up , boot the SDRLab again and see if you get the same error?

**Reply by vnegnev** [reply]
Another suggestion: if you still have it, you can use the original Red Pitaya SD card and try running `copy_bitstream.sh` with it. The commands to load the FPGA bitstream are different, but I'm curious whether that one will also fail.

**Reply by aaristotlerobert** [reply]
the following is my output when I ran `copy_bitstream.sh` after formatting the SD card:
`(base) electrical@electrical:~/marcos_extras$ ./copy_bitstream.sh 169.254.14.10 rp-122
Copying FPGA bitstream...
The authenticity of host '169.254.14.10 (169.254.14.10)' can't be established.
ECDSA key fingerprint is SHA256:TjO7BtBxp879H6RhhELatr2Pdms3q00sAvZCJmycgn8.
Are you sure you want to continue connecting (yes/no/[fingerprint])? yes
Warning: Permanently added '169.254.14.10' (ECDSA) to the list of known hosts.
ocra_mri_rp-122.bit.bin                       100% 3951KB  10.6MB/s   00:00    
ocra_mri_rp-122.dtbo                          100% 4751     2.3MB/s   00:00    
Writing bitstream to FPGA, ocra Linux image
(you should see a blue light appear) ...
Pseudo-terminal will not be allocated because stdin is not a terminal.
`
Is this line `Pseudo-terminal will not be allocated because stdin is not a terminal.` a right script if all is fine?
Still the blue light didn't appear

**Reply by vnegnev** [reply]
The `Pseudo-terminal will not be allocated...` line is a slight bug which I haven't fixed yet, but it doesn't indicate a problem. It means the debugging information from the commands run by `copy_bitstream.sh` on the SDRLab won't get printed by the script. That's okay if there's no problem with loading the bitstream, but we want to see their output.

Could you please reboot the SDRLab using the `poweroff` command, wait 10 seconds, then power off/on the supply? Then as soon as the SDRLab boots, run these commands again without running anything before:

```
root@redpitaya:~# echo 0 > /sys/class/fpga_manager/fpga0/flags
root@redpitaya:~# mkdir /sys/kernel/config/device-tree/overlays/full
root@redpitaya:~# echo -n "ocra_mri.dtbo" > /sys/kernel/config/device-tree/overlays/full/path
```

Also, just to double-check, are you using the original power supply for the SDRLab (or another one capable of 5V 2A)?

**Reply by aaristotlerobert** [reply]
Oh! Let me do that tomorrow as am already off the lab

**Reply by aaristotlerobert** [reply]
Hi Vlad,
I input the commands, as you had said above and everything seems ok because there was no error shown. So what should I do next. Thanks

**Reply by vnegnev** [reply]
That's a good sign, did you see the blue light come on? If it came on fine, then you can try running the main `marcos_setup.sh` script again and see if you can get sequences going on the server.

**Comment by LUMC-LowFieldMRI** [comment]
We just checked and for Robert the light didnt come on, with us in Leiden it did. This was tested with the RP plugged in to a 2.1 Amp, 5V power supply

**Reply by vnegnev** [reply]
Ok, that's a bit concerning - can you try using the original Red Pitaya SD card and make sure you can use that to program the FPGA (using the Red Pitaya instructions)?

**Reply by vnegnev** [reply]
I can also recommend setting up the serial debugging during boot-up - this will help find any problems. If you follow the steps here, maybe we will see what's going wrong:

https://redpitaya.readthedocs.io/en/latest/developerGuide/software/console/console/console.html

**Comment by josalggui** [comment]
Hi Vlad. Over the past few weeks, I've encountered difficulties with running `copy_bitstream.sh` while setting up MaRCoS from scratch. I've noticed that the blue light doesn't turn on during this process. We've tried using different SDRLab and SD cards, but the results remain identical. The output in the terminal after executing `copy_bitstream.sh` appears normal, without any error messages or warnings. The only method that seems to work is cloning a functioning SD card. Do you think there might be a new factor causing this issue?

**Reply by vnegnev** [reply]
I'll reference this in a new issue, since it probably isn't the same problem as before. Sorry about the delay.

**Comment by vnegnev** [comment]
I believe I've solved the issue with [this commit](https://github.com/vnegnev/marcos_extras/commit/4fd6c8862309cd40c85fe93d6f6e974cc757a017) - please feel free to test it, and comment in the [new issue](https://github.com/vnegnev/marcos_extras/issues/17) if there are any further similar problems observed.
